File: backend/src/services/ollamaService.ts
const OLLAMA_HOST = process.env.OLLAMA_HOST || 'http://localhost:11434';
const OLLAMA_MODEL = process.env.OLLAMA_MODEL || 'qwen2:1.5b';

export interface OllamaResponse {
  model: string;
  response: string;
  done: boolean;
}

const ALLOWED_BRIEF_TYPES = new Set(['world', 'security', 'political', 'humanitarian', 'military', 'economic']);

export async function generateBrief(headlines: string[], briefType: string = 'world'): Promise<string> {
  const safeType = ALLOWED_BRIEF_TYPES.has(briefType) ? briefType : 'world';
  const headlineList = headlines.map(h => `- ${h}`).join('\n');
  const prompt = `You are a geopolitical intelligence analyst. Given these recent headlines, write a concise 3-paragraph "${safeType.charAt(0).toUpperCase() + safeType.slice(1)} Brief" summarizing the most significant global developments, key themes, and any emerging risks.
Headlines:
${headlineList}
Write in the style of a professional intelligence briefing. Be factual and concise.`;

  const response = await fetch(`${OLLAMA_HOST}/api/generate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      model: OLLAMA_MODEL,
      prompt,
      stream: false,
    }),
    signal: AbortSignal.timeout(60000),
  });

  if (!response.ok) {
    throw new Error(`Ollama API error: ${response.status} ${response.statusText}`);
  }

  const data = await response.json() as OllamaResponse;
  return data.response;
}

export async function checkOllamaHealth(): Promise<boolean> {
  try {
    const response = await fetch(`${OLLAMA_HOST}/api/tags`, {
      signal: AbortSignal.timeout(5000),
    });
    return response.ok;
  } catch {
    return false;
  }
}

export { OLLAMA_MODEL };

File: backend/src/main/java/com/goldmonitor/service/OllamaService.java
package com.goldmonitor.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import java.util.List;
import java.util.Map;
import java.util.Set;

@Service
public class OllamaService {

    private static final Logger log = LoggerFactory.getLogger(OllamaService.class);
    private static final Set<String> ALLOWED_TYPES =
            Set.of("world", "security", "political", "humanitarian", "military", "economic");

    @Value("${ollama.host}")
    private String ollamaHost;

    @Value("${ollama.model}")
    private String ollamaModel;

    public String getModel() {
        return ollamaModel;
    }

    public boolean checkHealth() {
        try {
            RestClient client = RestClient.create();
            var response = client.get()
                    .uri(ollamaHost + "/api/tags")
                    .retrieve()
                    .toBodilessEntity();
            return response.getStatusCode().is2xxSuccessful();
        } catch (Exception e) {
            log.warn("Ollama health check failed: {}", e.getMessage());
            return false;
        }
    }

    @SuppressWarnings("unchecked")
    public String generateBrief(List<String> headlines, String briefType) {
        String safeType = ALLOWED_TYPES.contains(briefType) ? briefType : "world";
        StringBuilder headlineList = new StringBuilder();
        for (String h : headlines) {
            headlineList.append("- ").append(h).append("\n");
        }

        String capitalised = safeType.substring(0, 1).toUpperCase() + safeType.substring(1);
        String prompt = String.format(
                "You are a geopolitical intelligence analyst. Given these recent headlines, " +
                "write a concise 3-paragraph \"%s Brief\" summarizing the most significant global " +
                "developments, key themes, and any emerging risks.\nHeadlines:\n%s\n" +
                "Write in the style of a professional intelligence briefing. Be factual and concise.",
                capitalised, headlineList);

        RestClient client = RestClient.create();
        Map<String, Object> body = Map.of(
                "model", ollamaModel,
                "prompt", prompt,
                "stream", false);

        Map<String, Object> result = client.post()
                .uri(ollamaHost + "/api/generate")
                .header("Content-Type", "application/json")
                .body(body)
                .retrieve()
                .body(Map.class);

        if (result == null || !result.containsKey("response")) {
            throw new RuntimeException("Empty response from Ollama");
        }
        return (String) result.get("response");
    }
}
